presto-hive-metastore/src/main/java/com/facebook/presto/hive/TypeTranslator.java

```diff
@@ -16,7 +16,11 @@
 import com.facebook.presto.common.type.Type;
 import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
 
+import java.util.Optional;
+
 public interface TypeTranslator
 {
     TypeInfo translate(Type type);
+
+    TypeInfo translate(Type type, Optional<HiveType> defaultHiveType);
```

Comment:
I would simply change the signature of the first method to `Optional<TypeInfo> tryTranslate(Type type)`

Reply:
I want to keep the old behavior to throw an error inside translate function, so that 1.  all call sites don't need to check returned TypeInfo. 2,  HiveType's constructor method and `HiveType::toHiveType(TypeTranslator,  Type)` are not affected.  

DefaultHiveType is optional, caller will make decision on throwing an error or convert to default.

Reply:
How about adding `Optional<TypeInfo> tryTranslate(Type type)`? This signature might be more flexible, as it allows the caller to figure out what to do if the type mapping hasn't been satisfied.

Reply:
`translate` function calls itself a couple of times, handling of failed translation inside function is unavoidable.

Reply:
I'm slow. You are right, the method is recursive.

Reply:
Maybe add some comment explaining what is going on. The `defaultHiveType` might be really confusing.